### ISSUE TITLE
[Docs] SGPR's inducing points not be learned?

### ISSUE THREAD

**Issue by mxef** [cornellius-gp/gpytorch#1380]
I am trying to implement the [SGPR](https://docs.gpytorch.ai/en/v1.1.1/examples/02_Scalable_Exact_GPs/SGPR_Regression_CUDA.html) notebook. According to the [paper](http://proceedings.mlr.press/v5/titsias09a/titsias09a.pdf) and the notebook the inducing points locations are learned from the data.

I am using Two Patterns: a ucr_uea (time-series) dataset from the tslearn.datasets. I am trying to fit the first time-series (length 128) with the SGPR.  If I choose the inducing points to be the first 50 points of train_x, then I see that the locations of these points are not learned. I can see that from model.state_dict() and also from the plot (test_x = train_x).

Am I missing something here? Thank you!

``` python
import numpy as np
import matplotlib.pyplot as plt
from tslearn.datasets import UCR_UEA_datasets
import math
import torch
import gpytorch

# Make plots inline
%matplotlib inline

dataset = "TwoPatterns"

X_train, y_train, X_test, y_test = UCR_UEA_datasets().load_dataset(dataset)

train_y = X_train[0,:,0]

train_x = torch.from_numpy(np.arange(1,train_y.shape[0]+1)).float()
train_y = torch.from_numpy(train_y).float()

train_y = train_y - torch.min(train_y)
train_y = 2 * (train_y / torch.max(train_y)) - 1

from gpytorch.means import ConstantMean
from gpytorch.kernels import ScaleKernel, RBFKernel, InducingPointKernel
from gpytorch.distributions import MultivariateNormal

class GPRegressionModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(GPRegressionModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = ConstantMean()
        self.base_covar_module = ScaleKernel(RBFKernel())   
        self.covar_module = InducingPointKernel(self.base_covar_module, inducing_points=train_x[:50]
, likelihood=likelihood)

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultivariateNormal(mean_x, covar_x)

likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = GPRegressionModel(train_x, train_y, likelihood)

training_iterations = 50

# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.1)

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

def train():
    for i in range(training_iterations):
        # Zero backprop gradients
        optimizer.zero_grad()
        # Get output from model
        output = model(train_x)
        # Calc loss and backprop derivatives
        loss = -mll(output, train_y)
        loss.backward()
        print('Iter %d/%d - Loss: %.3f' % (i + 1, training_iterations, loss.item()))
        optimizer.step()

# See dkl_mnist.ipynb for explanation of this flag
%time train()


model.state_dict()


model.eval()
likelihood.eval()
with gpytorch.settings.max_preconditioner_size(10), torch.no_grad():
    with gpytorch.settings.max_root_decomposition_size(30), gpytorch.settings.fast_pred_var():
        preds = model(train_x)
        
with torch.no_grad():
        # Initialize plot
        f, ax = plt.subplots(1, 1, figsize=(12, 9))

        # Get upper and lower confidence bounds
        lower, upper = preds.confidence_region()
        # Plot training data as black stars
        ax.plot(train_x.numpy(), train_y.numpy(), 'k*')
        # Plot predictive means as blue line
        ax.plot(train_x.numpy(), preds.mean.numpy(), 'b')
        # Shade between the lower and upper confidence bounds
        ax.fill_between(train_x.numpy(), lower.numpy(), upper.numpy(), alpha=0.5)
        ax.set_ylim([-3, 3])
        ax.legend(['Observed Data', 'Mean', 'Confidence'])
        plt.show()
        print('Test MAE: {}'.format(torch.mean(torch.abs(preds.mean - train_y))))
```

**Comment by jacobrgardner** [comment]
Best I can tell this is just an initialization and data scale thing.

Adding feature normalization:
```diff
train_x = torch.from_numpy(np.arange(1,train_y.shape[0]+1)).float()
+ train_x = (train_x - train_x.mean(0)) / train_x.std(0)

train_y = torch.from_numpy(train_y).float()
```
and just doing random inducing point initialization:
```diff
- self.covar_module = InducingPointKernel(self.base_covar_module, inducing_points=train_x[:50]
+ self.covar_module = InducingPointKernel(self.base_covar_module, inducing_points=torch.randn(50)
```
results both in the inducing points being substantially different from the initialization and a much better data fit:
![image](https://user-images.githubusercontent.com/4016393/101406172-1a7c9c00-38a7-11eb-8946-011135f95f6e.png)

In general, I'd recommend that you always normalize both your features and labels when using GPyTorch. Some of our hyperparameter initialization basically assumes that your data has distances on the order of 1.

**Comment by mxef** [comment]
I ran it, with the changes, multiple times and I get this:
![download](https://user-images.githubusercontent.com/45259486/101409202-3d7a7080-38e6-11eb-8054-12512c502145.png)

**Comment by jacobrgardner** [comment]
Here's my full code that got the plot above:

```python
import numpy as np
import matplotlib.pyplot as plt
from tslearn.datasets import UCR_UEA_datasets
import math
import torch
import gpytorch

# Make plots inline
%matplotlib inline

dataset = "TwoPatterns"

X_train, y_train, X_test, y_test = UCR_UEA_datasets().load_dataset(dataset)

train_y = X_train[0,:,0]

train_x = torch.from_numpy(np.arange(1,train_y.shape[0]+1)).float()
train_x = (train_x - train_x.mean(0)) / train_x.std(0)

train_y = torch.from_numpy(train_y).float()

train_y = train_y - torch.min(train_y)
train_y = 2 * (train_y / torch.max(train_y)) - 1

from gpytorch.means import ConstantMean
from gpytorch.kernels import ScaleKernel, RBFKernel, InducingPointKernel
from gpytorch.distributions import MultivariateNormal

class GPRegressionModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(GPRegressionModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = ConstantMean()
        self.base_covar_module = ScaleKernel(RBFKernel())
        self.covar_module = InducingPointKernel(self.base_covar_module, inducing_points=torch.randn(50)
, likelihood=likelihood)

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultivariateNormal(mean_x, covar_x)

likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = GPRegressionModel(train_x, train_y, likelihood)

print(model.covar_module.inducing_points)

training_iterations = 50

# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam(model.parameters(), lr=0.1)

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

def train():
    for i in range(training_iterations):
        # Zero backprop gradients
        optimizer.zero_grad()
        # Get output from model
        output = model(train_x)
        # Calc loss and backprop derivatives
        loss = -mll(output, train_y)
        loss.backward()
#         print('Iter %d/%d - Loss: %.3f' % (i + 1, training_iterations, loss.item()))
        optimizer.step()

# See dkl_mnist.ipynb for explanation of this flag
%time train()


model.state_dict()


model.eval()
likelihood.eval()
with gpytorch.settings.max_preconditioner_size(10), torch.no_grad():
    with gpytorch.settings.max_root_decomposition_size(30), gpytorch.settings.fast_pred_var():
        preds = model(train_x)
        
with torch.no_grad():
        # Initialize plot
        f, ax = plt.subplots(1, 1, figsize=(12, 9))

        # Get upper and lower confidence bounds
        lower, upper = preds.confidence_region()
        # Plot training data as black stars
        ax.plot(train_x.numpy(), train_y.numpy(), 'k*')
        # Plot predictive means as blue line
        ax.plot(train_x.numpy(), preds.mean.numpy(), 'b')
        # Shade between the lower and upper confidence bounds
        ax.fill_between(train_x.numpy(), lower.numpy(), upper.numpy(), alpha=0.5)
        ax.set_ylim([-3, 3])
        ax.legend(['Observed Data', 'Mean', 'Confidence'])
        plt.show()
#         print('Test MAE: {}'.format(torch.mean(torch.abs(preds.mean - train_y))))
```

**Comment by mxef** [comment]
Ok, now it works! The problem was that I used the **torch.optim.SGD** instead of **torch.optim.Adam**. I think you should change the following in [SGPR tutorial](https://docs.gpytorch.ai/en/v1.1.1/examples/02_Scalable_Exact_GPs/SGPR_Regression_CUDA.html):
``` python
# Use the adam optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
```
Thanks again!

**Comment by jacobrgardner** [comment]
Ah sorry, switching SGD to Adam for our GP models is such a knee jerk reaction for me that I didn't even notice making it when I made my first response. We should definitely change it in the tutorial

**Comment by gpleiss** [comment]
I'm putting up a PR to change SGD to Adam in all of our examples/tutorials. Closing this issue though.
